Signal/src/ViewControllers/ConversationView/ConversationViewItem.m

```diff
@@ -266,49 +274,71 @@ - (NSCache *)displayableTextCache
     return cache;
 }
 
-- (NSString *)displayableTextForText:(NSString *)text interactionId:(NSString *)interactionId
+- (DisplayableText *)displayableTextForText:(NSString *)text interactionId:(NSString *)interactionId
 {
     OWSAssert(text);
     OWSAssert(interactionId.length > 0);
 
-    NSString *_Nullable displayableText = [[self displayableTextCache] objectForKey:interactionId];
+    return [self displayableTextForInteractionId:interactionId
+                                       textBlock:^{
+                                           return text;
+                                       }];
+}
+
+- (DisplayableText *)displayableTextForAttachmentStream:(TSAttachmentStream *)attachmentStream
+                                          interactionId:(NSString *)interactionId
+{
+    OWSAssert(attachmentStream);
+    OWSAssert(interactionId.length > 0);
+
+    return [self displayableTextForInteractionId:interactionId
+                                       textBlock:^{
+                                           NSData *textData = [NSData dataWithContentsOfURL:attachmentStream.mediaURL];
+                                           NSString *text =
+                                               [[NSString alloc] initWithData:textData encoding:NSUTF8StringEncoding];
+                                           return text;
+                                       }];
+}
+
+- (DisplayableText *)displayableTextForInteractionId:(NSString *)interactionId
+                                           textBlock:(NSString * (^_Nonnull)())textBlock
+{
+    OWSAssert(interactionId.length > 0);
+
+    DisplayableText *_Nullable displayableText = [[self displayableTextCache] objectForKey:interactionId];
     if (!displayableText) {
+        NSString *text = textBlock();
+
         // Only show up to 2kb of text.
         const NSUInteger kMaxTextDisplayLength = 2 * 1024;
```

Comment:
I've unified the "strip, de-zalgo and possibly truncate" logic here.

Reply:
This also will cut down on a bunch of redundant (and possibly expensive) work by leaning on the cache more.